File: scripts/scrape.mjs
import * as cheerio from 'cheerio';
import { writeFileSync, mkdirSync, existsSync } from 'fs';
import { join } from 'path';
import https from 'https';

// All categories from cgs.hk navigation
const CATEGORIES = [
  { id: 'sword', url: 'https://cgs.hk/produce1.htm', label: '鑄劍' },
  { id: 'axe', url: 'https://cgs.hk/produce2.htm', label: '造斧' },
  { id: 'spear', url: 'https://cgs.hk/produce3.htm', label: '造槍' },
  { id: 'bow', url: 'https://cgs.hk/produce4.htm', label: '造弓' },
  { id: 'staff', url: 'https://cgs.hk/produce5.htm', label: '造杖' },
  { id: 'dagger', url: 'https://cgs.hk/produce6.htm', label: '小刀' },
  { id: 'throw', url: 'https://cgs.hk/produce7.htm', label: '投擲武器' },
  { id: 'bomb', url: 'https://cgs.hk/produce8.htm', label: '炸彈' },
  { id: 'helmet', url: 'https://cgs.hk/produce9.htm', label: '頭盔' },
  { id: 'hat', url: 'https://cgs.hk/produce10.htm', label: '帽子' },
  { id: 'armor', url: 'https://cgs.hk/produce11.htm', label: '鎧甲' },
  { id: 'cloth', url: 'https://cgs.hk/produce12.htm', label: '衣服' },
  { id: 'robe', url: 'https://cgs.hk/produce13.htm', label: '長袍' },
  { id: 'boots', url: 'https://cgs.hk/produce14.htm', label: '靴子' },
  { id: 'shoes', url: 'https://cgs.hk/produce15.htm', label: '鞋子' },
  { id: 'shield', url: 'https://cgs.hk/produce16.htm', label: '盾牌' },
  { id: 'cooking', url: 'https://cgs.hk/produce17.htm', label: '料理' },
  { id: 'pharmacy', url: 'https://cgs.hk/produce18.htm', label: '藥品' },
  { id: 'accessory', url: 'https://cgs.hk/produce19.htm', label: '飾品' },
  { id: 'dragon', url: 'https://cgs.hk/produce20.htm', label: '水龍系列' },
  { id: 'fiveC', url: 'https://cgs.hk/produce21.htm', label: '５Ｃ系列' },
  { id: 'scroll', url: 'https://cgs.hk/produce27.htm', label: '變身卷軸' },
  { id: 'collar', url: 'https://cgs.hk/produce22.htm', label: '寵物項圈' },
  { id: 'crystal', url: 'https://cgs.hk/produce23.htm', label: '寵物晶石' },
  { id: 'petArmor', url: 'https://cgs.hk/produce24.htm', label: '寵物裝甲' },
  { id: 'petAccessory', url: 'https://cgs.hk/produce25.htm', label: '寵物飾品' },
  { id: 'petCloth', url: 'https://cgs.hk/produce26.htm', label: '寵物服裝' },
];

function fetchPage(url) {
  return new Promise((resolve, reject) => {
    https.get(url, res => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
      res.on('error', reject);
    }).on('error', reject);
  });
}

function parseLevelVariant(text) {
  // e.g. "1A", "2B", "13C"
  const match = text.match(/^(\d+)([A-Z])$/);
  if (!match) return null;
  return { level: parseInt(match[1]), variant: match[2] };
}

function parseMaterialLevel(title) {
  // e.g. "等級1", "等級12"
  if (!title) return 0;
  const match = title.match(/等級(\d+)/);
  return match ? parseInt(match[1]) : 0;
}

function parsePage(html, categoryId) {
  const $ = cheerio.load(html);
  const recipes = [];

  $('tbody[id]').each((_, tbody) => {
    const rows = $(tbody).find('> tr').toArray();
    if (rows.length < 3) return;

    const row1 = $(rows[0]); // images + quantities
    const row2 = $(rows[1]); // names
    const row3 = $(rows[2]); // type + data-set values

    // Level/variant from first cell
    const lvText = row1.find('td').first().text().trim();
    const parsed = parseLevelVariant(lvText);
    if (!parsed) return;

    const { level, variant } = parsed;
    const id = `${categoryId}-${level}${variant.toLowerCase()}`;

    // Row2 cells: [itemName, mat1Name, mat2Name, ...]
    // (row1's first td has rowspan=3, so row2 doesn't have the level cell)
    const row2Cells = row2.find('td').toArray();
    const itemName = $(row2Cells[0]).text().trim();
    if (!itemName) return;

    // Item image (second cell in row1, index 1)
    const itemImgCell = row1.find('td').eq(1);
    const itemImg = itemImgCell.find('img').attr('data-src') || '';

    // Stats from the last cell with rowspan=3 in row1
    const allCellsRow1 = row1.find('td').toArray();
    const lastCell = $(allCellsRow1[allCellsRow1.length - 1]);
    const stats = lastCell.attr('rowspan') ? lastCell.text().trim().replace(/\s+/g, ' ') : '';

    // MP cost = level * 20
    const mpCost = level * 20;

    // Materials: quantity cells in row1 (class="quantity")
    const materials = [];
    let matIdx = 0;
    for (let i = 2; i < allCellsRow1.length; i++) {
      const cell1 = $(allCellsRow1[i]);
      if (cell1.attr('rowspan')) continue; // skip stats column

      const img = cell1.find('img');
      if (!img.length) { matIdx++; continue; }

      const matImg = img.attr('data-src') || '';
      const matLevel = parseMaterialLevel(img.attr('title'));
      const qtyText = cell1.find('span').text().trim();
      const qtyMatch = qtyText.match(/x\s*(\d+)/);
      const quantity = qtyMatch ? parseInt(qtyMatch[1]) : 0;

      // Material name from row2: cells[1+matIdx]
      const nameCell = row2Cells[1 + matIdx];
      const matName = nameCell ? $(nameCell).text().trim() : '';

      matIdx++;

      if (matName && quantity > 0) {
        materials.push({
          name: matName,
          quantity,
          image: matImg + '.png',
          materialLevel: matLevel,
        });
      }
    }

    if (materials.length === 0) return;

    recipes.push({
      id,
      name: itemName,
      category: categoryId,
      level,
      variant,
      image: itemImg + '.png',
      mpCost,
      stats,
      materials,
    });
  });

  return recipes;
}

async function downloadImage(hash, outputDir) {
  const filename = hash + '.png';
  const filepath = join(outputDir, filename);
  if (existsSync(filepath)) return;

  const prefix = hash.substring(0, 2);
  const url = `https://ig2.cgs.hk/g/${prefix}/${hash}.png`;

  return new Promise((resolve) => {
    https.get(url, res => {
      if (res.statusCode !== 200) {
        console.warn(`  Failed to download ${hash}: ${res.statusCode}`);
        resolve();
        return;
      }
      const chunks = [];
      res.on('data', chunk => chunks.push(chunk));
      res.on('end', () => {
        const buffer = Buffer.concat(chunks);
        writeFileSync(filepath, buffer);
        resolve();
      });
      res.on('error', () => resolve());
    }).on('error', () => resolve());
  });
}

async function main() {
  const dataDir = join(process.cwd(), 'src/data');
  const imgDir = join(process.cwd(), 'public/items');
  mkdirSync(dataDir, { recursive: true });
  mkdirSync(imgDir, { recursive: true });

  const allCategories = [];

  for (const cat of CATEGORIES) {
    console.log(`Fetching ${cat.label} (${cat.url})...`);
    try {
      const html = await fetchPage(cat.url);
      const recipes = parsePage(html, cat.id);
      console.log(`  Found ${recipes.length} recipes`);

      if (recipes.length === 0) continue;

      // Save JSON
      const jsonPath = join(dataDir, `${cat.id}.json`);
      writeFileSync(jsonPath, JSON.stringify(recipes, null, 2));

      // Collect all image hashes to download
      const hashes = new Set();
      for (const recipe of recipes) {
        hashes.add(recipe.image.replace('.png', ''));
        for (const mat of recipe.materials) {
          hashes.add(mat.image.replace('.png', ''));
        }
      }

      console.log(`  Downloading ${hashes.size} images...`);
      const batchSize = 10;
      const hashArray = [...hashes];
      for (let i = 0; i < hashArray.length; i += batchSize) {
        const batch = hashArray.slice(i, i + batchSize);
        await Promise.all(batch.map(h => downloadImage(h, imgDir)));
      }

      allCategories.push({ id: cat.id, label: cat.label, count: recipes.length });
    } catch (err) {
      console.error(`  Error: ${err.message}`);
    }

    // Small delay between pages
    await new Promise(r => setTimeout(r, 500));
  }

  console.log('\n=== Summary ===');
  for (const cat of allCategories) {
    console.log(`${cat.label} (${cat.id}): ${cat.count} recipes`);
  }
}

main().catch(console.error);

File: scripts/scrape.py
#!/usr/bin/env python3
"""Scrape recipe data and item images from cgs.hk produce pages."""

import json
import os
import re
import sys
from pathlib import Path

import requests
from bs4 import BeautifulSoup

BASE_DIR = Path(__file__).resolve().parent.parent
DATA_DIR = BASE_DIR / "src" / "data"
ITEMS_DIR = BASE_DIR / "public" / "items"

PAGES = {
    "bow": "https://cgs.hk/produce4.htm",
    "cooking": "https://cgs.hk/produce17.htm",
}

SESSION = requests.Session()
SESSION.headers.update({"User-Agent": "Mozilla/5.0"})

downloaded_images: set[str] = set()


def fetch_page(url: str) -> BeautifulSoup:
    resp = SESSION.get(url, timeout=30)
    resp.raise_for_status()
    resp.encoding = resp.apparent_encoding
    return BeautifulSoup(resp.text, "html.parser")


def get_image_base(soup: BeautifulSoup) -> str:
    meta = soup.find("meta", attrs={"name": "igbase"})
    if meta and meta.get("content"):
        return meta["content"].strip()
    return "ig2.cgs.hk"


def image_url(base: str, hash_val: str) -> str:
    return f"https://{base}/g/{hash_val[:2]}/{hash_val}.png"


def download_image(base: str, hash_val: str) -> None:
    if not hash_val or hash_val in downloaded_images:
        return
    downloaded_images.add(hash_val)
    dest = ITEMS_DIR / f"{hash_val}.png"
    if dest.exists():
        return
    url = image_url(base, hash_val)
    try:
        resp = SESSION.get(url, timeout=15)
        resp.raise_for_status()
        dest.write_bytes(resp.content)
        print(f"  Downloaded {hash_val}.png")
    except Exception as e:
        print(f"  WARNING: Failed to download {hash_val}.png: {e}")


def parse_material_level(img_tag) -> int:
    """Extract material level from title='等級N'."""
    title = img_tag.get("title", "") if img_tag else ""
    m = re.search(r"等級(\d+)", title)
    return int(m.group(1)) if m else 1


def parse_variant(text: str):
    """Parse variant text like '1A' into (level, variant_letter)."""
    text = text.strip()
    m = re.match(r"(\d+)\s*([A-Za-z])", text)
    if m:
        return int(m.group(1)), m.group(2).upper()
    # fallback: try just a number
    m2 = re.match(r"(\d+)", text)
    if m2:
        return int(m2.group(1)), "A"
    return 1, "A"


def parse_quantity(td) -> int | None:
    """Extract quantity from a <td class='quantity'> cell."""
    span = td.find("span")
    if span:
        m = re.search(r"x\s*(\d+)", span.get_text())
        if m:
            return int(m.group(1))
    return None


def scrape_category(category: str, url: str) -> list[dict]:
    print(f"Scraping {category} from {url}")
    soup = fetch_page(url)
    ig_base = get_image_base(soup)
    print(f"  Image base: {ig_base}")

    recipes = []
    tbodies = soup.find_all("tbody", id=re.compile(r"^\d+$"))
    print(f"  Found {len(tbodies)} tbody blocks")

    for tbody in tbodies:
        try:
            rows = tbody.find_all("tr", recursive=False)
            # Filter out sub rows
            main_rows = [r for r in rows if "sub" not in (r.get("class") or [])]
            if len(main_rows) < 3:
                continue

            row1, row2, row3 = main_rows[0], main_rows[1], main_rows[2]

            # Row 1: variant cell (rowspan=3), product image, materials
            row1_tds = row1.find_all("td", recursive=False)
            if len(row1_tds) < 2:
                continue

            # First td is variant
            variant_text = row1_tds[0].get_text(strip=True)
            level, variant_letter = parse_variant(variant_text)

            # Second td has product image
            product_img_tag = row1_tds[1].find("img")
            product_hash = ""
            if product_img_tag:
                product_hash = product_img_tag.get("data-src", "").strip()

            # Material tds (class="quantity")
            mat_tds_row1 = [td for td in row1_tds[2:] if "quantity" in (td.get("class") or [])]

            # Row 2: names (product name, then material names)
            row2_tds = row2.find_all("td", recursive=False)

            # Row 3: type + data-set values
            row3_tds = row3.find_all("td", recursive=False)

            # Product name from row2 first td
            product_name = row2_tds[0].get_text(strip=True) if row2_tds else ""

            # Build materials list
            materials = []
            for i, mat_td in enumerate(mat_tds_row1):
                qty = parse_quantity(mat_td)
                if qty is None:
                    continue

                mat_img = mat_td.find("img")
                mat_hash = mat_img.get("data-src", "").strip() if mat_img else ""
                mat_level = parse_material_level(mat_img)

                # Material name from row2 (offset by 1 for product name)
                mat_name_idx = i + 1
                mat_name = ""
                if mat_name_idx < len(row2_tds):
                    mat_name = row2_tds[mat_name_idx].get_text(strip=True)

                # Base quantity from row3 (offset by 1 for type)
                base_qty_idx = i + 1
                base_qty = qty
                if base_qty_idx < len(row3_tds):
                    bq_text = row3_tds[base_qty_idx].get_text(strip=True)
                    if bq_text.isdigit():
                        base_qty = int(bq_text)

                if mat_hash:
                    download_image(ig_base, mat_hash)

                materials.append({
                    "name": mat_name,
                    "quantity": base_qty,
                    "image": f"{mat_hash}.png" if mat_hash else "",
                    "materialLevel": mat_level,
                })

            if product_hash:
                download_image(ig_base, product_hash)

            recipe_id = f"{category}-{level}{variant_letter.lower()}"
            recipe = {
                "id": recipe_id,
                "name": product_name,
                "category": category,
                "level": level,
                "variant": variant_letter,
                "image": f"{product_hash}.png" if product_hash else "",
                "mpCost": level * 20,
                "materials": materials,
            }
            recipes.append(recipe)

        except Exception as e:
            tbody_id = tbody.get("id", "?")
            print(f"  WARNING: Failed to parse tbody id={tbody_id}: {e}")
            continue

    print(f"  Parsed {len(recipes)} recipes")
    return recipes


def main():
    DATA_DIR.mkdir(parents=True, exist_ok=True)
    ITEMS_DIR.mkdir(parents=True, exist_ok=True)

    for category, url in PAGES.items():
        recipes = scrape_category(category, url)
        out_path = DATA_DIR / f"{category}.json"
        with open(out_path, "w", encoding="utf-8") as f:
            json.dump(recipes, f, ensure_ascii=False, indent=2)
        print(f"  Saved {len(recipes)} recipes to {out_path}")

    print(f"\nTotal images downloaded: {len(downloaded_images)}")
    print("Done!")


if __name__ == "__main__":
    main()
